scripts/convert-to-safari-rules.ts

```diff
@@ -1,14 +1,276 @@
+      isOfType(elem, 'string', 'domain must be a string');
+      //expect(elem === elem.toLowerCase()).toBe(true);
+      isSame(elem, elem.toLowerCase(), 'domain must be a lowercase string');
+      // TODO: test if string is punycode encoded
+    }
+  }
+
+  // type must be a string with the following possible values
+  value = rule.action.type;
+  //expect(allowedActions.has(value)).toBe(true);
+  isSame(allowedActions.has(value), true, 'invalid type');
+
+  // if type is css-display-none, a non-empty selector must exist
+  if (rule.action.type === 'css-display-none') {
+    //expect(rule.action.selector).not.toBeUndefined();
+    isNotSame(rule.action.selector, undefined, 'if type is css-display-none, a non-empty selector must exist');
+    //expect(rule.action.selector).not.toBe('');
+    isNotSame(rule.action.selector, '', 'selector cannot be an empty string when the type is css-display-none');
+  }
+
+  // selector must be a non-empty string
+  value = rule.action.selector;
+  if (value !== undefined) {
+    //expect(typeof value).toBe('string');
+    isOfType(value, 'string', 'selector must be a string');
+    //expect(value).not.toBe('');
+    isNotSame(value, '', 'selector cannot be an emtpy string');
+  }
+
+  // trigger cannot contain both if-domain and unless-domain
+  const ifdomain = rule.trigger['if-domain'];
+  const unlessdomain = rule.trigger['unless-domain'];
+  if (ifdomain !== undefined) {
+    //expect(unlessdomain).toBeUndefined();
+    isSame(unlessdomain, undefined, 'trigger cannot contain both if-domain and unless-domain');
+  } else if (unlessdomain !== undefined) {
+    //expect(ifdomain).toBeUndefined();
+    isSame(ifdomain, undefined, 'trigger cannot contain both if-domain and unless-domain');
+  }
+
+  // trigger cannot contain both if-top-url and unless-top-url
+  const ifurl = rule.trigger['if-top-url'];
+  const unlessurl = rule.trigger['unless-top-url'];
+  if (ifurl !== undefined) {
+    //expect(unlessurl).toBeUndefined();
+    isSame(unlessurl, undefined, 'trigger cannot contain both if-top-url and unless-top-url');
+  } else if (unlessurl !== undefined) {
+    //expect(ifurl).toBeUndefined();
+    isSame(ifurl, undefined, 'trigger cannot contain both if-top-url and unless-top-url');
+  }
+}
+
+export function convertAndValidateFilters(lists) {
+  const { networkFilters, cosmeticFilters } = parseList(lists);
+  let json: any[] = [];
+  for (let i = 0; i < cosmeticFilters.length; i += 1) {
+    const cosmetic = cosmeticFilters[i];
+    const rule = convertCosmetics(cosmetic);
+    if (rule !== null && rule !== undefined) {
+      try {
+        testRule(rule);
+      } catch (ex) {
+        return [];
+      }
+      json.push(rule);
+    }
+  }
+
+  for (let j = 0; j < networkFilters.length; j += 1) {
+    const filter = networkFilters[j];
+    const rule = convertFilter(filter);
+    if (rule !== null && rule !== undefined) {
+      try {
+        testRule(rule);
+      } catch (ex) {
+        return [];
```

Comment:
Not sure if catching the exception and returning empty list is something you would expect from this function? Why not letting the exception go up?